service/history/historyEngine.go

```diff
@@ -690,108 +698,173 @@ func (e *historyEngineImpl) QueryWorkflow(
 	ctx ctx.Context,
 	request *h.QueryWorkflowRequest,
 ) (retResp *h.QueryWorkflowResponse, retErr error) {
-	domainCache, err := e.shard.GetDomainCache().GetDomainByID(request.GetDomainUUID())
+
+	context, release, err := e.historyCache.getOrCreateWorkflowExecution(ctx, request.GetDomainUUID(), *request.GetRequest().GetExecution())
 	if err != nil {
 		return nil, err
 	}
-	context, release, err := e.historyCache.getOrCreateWorkflowExecution(ctx, request.GetDomainUUID(), *request.GetExecution())
+	defer release(retErr)
+	ms, err := context.loadWorkflowExecution()
 	if err != nil {
 		return nil, err
 	}
-	msBuilder, err := context.loadWorkflowExecution()
+	req := request.GetRequest()
+	if !ms.IsWorkflowExecutionRunning() && req.QueryRejectCondition != nil {
+		notOpenReject := req.GetQueryRejectCondition() == workflow.QueryRejectConditionNotOpen
+		_, closeStatus := ms.GetWorkflowStateCloseStatus()
+		notCompletedCleanlyReject := req.GetQueryRejectCondition() == workflow.QueryRejectConditionNotCompletedCleanly && closeStatus != persistence.WorkflowCloseStatusCompleted
+		if notOpenReject || notCompletedCleanlyReject {
+			return &h.QueryWorkflowResponse{
+				Response: &workflow.QueryWorkflowResponse{
+					QueryRejected: &workflow.QueryRejected{
+						CloseStatus: persistence.ToThriftWorkflowExecutionCloseStatus(closeStatus).Ptr(),
+					},
+				},
+			}, nil
+		}
+	}
+
+	execInfo := ms.GetExecutionInfo()
+	clientFeature := client.NewFeatureImpl(execInfo.ClientLibraryVersion, execInfo.ClientFeatureVersion, execInfo.ClientImpl)
+
+	de, err := e.shard.GetDomainCache().GetDomainByID(request.GetDomainUUID())
 	if err != nil {
-		release(err)
 		return nil, err
 	}
-	queryRegistry := msBuilder.GetQueryRegistry()
-	release(nil)
-	queryID, _, queryTermCh := queryRegistry.bufferQuery(request.GetQuery())
-
-	ttl := e.shard.GetConfig().LongPollExpirationInterval(domainCache.GetInfo().Name)
-	timer := time.NewTimer(ttl)
 
-	// after query is finished removed query from query registry and remove any potentially created in memory decision task
-	defer func() {
-		timer.Stop()
-		queryRegistry.removeQuery(queryID)
-		context, release, err := e.historyCache.getOrCreateWorkflowExecution(ctx, request.GetDomainUUID(), *request.GetExecution())
-		if err != nil {
-			retResp = nil
-			retErr = err
-			return
-		}
-		msBuilder, err := context.loadWorkflowExecution()
-		if err != nil {
-			release(err)
-			retResp = nil
-			retErr = err
-			return
-		}
-		msBuilder.DeleteInMemoryDecisionTask()
+	scope := e.metricsClient.Scope(metrics.HistoryQueryWorkflowScope)
```

Comment:
metricsScope ?

Reply:
In other places here we we do the following 

```e.metricsClient.IncCounter(metrics.HistoryRespondActivityTaskFailedScope, metrics.StaleMutableStateCounter)```

Instead of redeclaring the scope every time a metric is emitted I think a better pattern is to create a scope object and use that to emit metrics.